### ISSUE TITLE
Repeated results on completion with progcomp

### ISSUE THREAD

**Issue by 3ximus** [akinomyoga/ble.sh#348]
GNU bash, version 5.2.15(1)-release (x86_64-pc-linux-gnu) [Ubuntu 23.04]
ble.sh, version 0.4.0-devel4+4089c4e1 (noarch) [git 2.39.2, GNU Make 4.3, GNU Awk 5.2.1, API 3.2, PMA Avon 8-g1, (GNU MPFR 4.2.0, GNU MP 6.2.1)]
bash-completion, version 2.11 (hash:8fb067d157b63461693a72b0885cd9b7f3c1490f, 77086 bytes) (noarch)
fzf key-bindings, (hash:d8ce533b85943bd535fe6ad487090e33f0f68b13, 3083 bytes) (noarch)
WARNING: fzf integration "integration/fzf-key-bindings" is not activated.
locale: LANG=en_CA.UTF-8
terminal: TERM=screen-256color wcwidth=15.0-west/15.0-2+ri, tmux:0 (84;0;0)

Hi there!
I've been noticing this issue with completion. I'm working on a small cli application that generates a completion file. However I'm getting duplicate completions.
I'm guessing it's likely because each completion has a description associated with it and so blesh is considering them as different? For example if I remove the description then only the `progcomp` entries are displayed and no duplicates exist.

I've attached a gif where on the top window I'm running `blesh` and on the bottom I've detached blesh and as you can see on the top the progcomp besides being duplicate entries insert the description text as well.

Do you have any clue what's causing it ? I'm also going to attache the generated bash completion file in case it helps. 

![Peek 2023-07-23 15-05](https://github.com/akinomyoga/ble.sh/assets/9083012/3db2f82f-beb4-42a5-a5c0-d73cad67da3b)

<details><summary>Completion Script</summary>
<p>

```bash
# bash completion V2 for bb                                   -*- shell-script -*-

__bb_debug()
{
    if [[ -n ${BASH_COMP_DEBUG_FILE-} ]]; then
        echo "$*" >> "${BASH_COMP_DEBUG_FILE}"
    fi
}

# Macs have bash3 for which the bash-completion package doesn't include
# _init_completion. This is a minimal version of that function.
__bb_init_completion()
{
    COMPREPLY=()
    _get_comp_words_by_ref "$@" cur prev words cword
}

# This function calls the bb program to obtain the completion
# results and the directive.  It fills the 'out' and 'directive' vars.
__bb_get_completion_results() {
    local requestComp lastParam lastChar args

    # Prepare the command to request completions for the program.
    # Calling ${words[0]} instead of directly bb allows to handle aliases
    args=("${words[@]:1}")
    requestComp="${words[0]} __complete ${args[*]}"

    lastParam=${words[$((${#words[@]}-1))]}
    lastChar=${lastParam:$((${#lastParam}-1)):1}
    __bb_debug "lastParam ${lastParam}, lastChar ${lastChar}"

    if [[ -z ${cur} && ${lastChar} != = ]]; then
        # If the last parameter is complete (there is a space following it)
        # We add an extra empty parameter so we can indicate this to the go method.
        __bb_debug "Adding extra empty parameter"
        requestComp="${requestComp} ''"
    fi

    # When completing a flag with an = (e.g., bb -n=<TAB>)
    # bash focuses on the part after the =, so we need to remove
    # the flag part from $cur
    if [[ ${cur} == -*=* ]]; then
        cur="${cur#*=}"
    fi

    __bb_debug "Calling ${requestComp}"
    # Use eval to handle any environment variables and such
    out=$(eval "${requestComp}" 2>/dev/null)

    # Extract the directive integer at the very end of the output following a colon (:)
    directive=${out##*:}
    # Remove the directive
    out=${out%:*}
    if [[ ${directive} == "${out}" ]]; then
        # There is not directive specified
        directive=0
    fi
    __bb_debug "The completion directive is: ${directive}"
    __bb_debug "The completions are: ${out}"
}

__bb_process_completion_results() {
    local shellCompDirectiveError=1
    local shellCompDirectiveNoSpace=2
    local shellCompDirectiveNoFileComp=4
    local shellCompDirectiveFilterFileExt=8
    local shellCompDirectiveFilterDirs=16
    local shellCompDirectiveKeepOrder=32

    if (((directive & shellCompDirectiveError) != 0)); then
        # Error code.  No completion.
        __bb_debug "Received error from custom completion go code"
        return
    else
        if (((directive & shellCompDirectiveNoSpace) != 0)); then
            if [[ $(type -t compopt) == builtin ]]; then
                __bb_debug "Activating no space"
                compopt -o nospace
            else
                __bb_debug "No space directive not supported in this version of bash"
            fi
        fi
        if (((directive & shellCompDirectiveKeepOrder) != 0)); then
            if [[ $(type -t compopt) == builtin ]]; then
                # no sort isn't supported for bash less than < 4.4
                if [[ ${BASH_VERSINFO[0]} -lt 4 || ( ${BASH_VERSINFO[0]} -eq 4 && ${BASH_VERSINFO[1]} -lt 4 ) ]]; then
                    __bb_debug "No sort directive not supported in this version of bash"
                else
                    __bb_debug "Activating keep order"
                    compopt -o nosort
                fi
            else
                __bb_debug "No sort directive not supported in this version of bash"
            fi
        fi
        if (((directive & shellCompDirectiveNoFileComp) != 0)); then
            if [[ $(type -t compopt) == builtin ]]; then
                __bb_debug "Activating no file completion"
                compopt +o default
            else
                __bb_debug "No file completion directive not supported in this version of bash"
            fi
        fi
    fi

    # Separate activeHelp from normal completions
    local completions=()
    local activeHelp=()
    __bb_extract_activeHelp

    if (((directive & shellCompDirectiveFilterFileExt) != 0)); then
        # File extension filtering
        local fullFilter filter filteringCmd

        # Do not use quotes around the $completions variable or else newline
        # characters will be kept.
        for filter in ${completions[*]}; do
            fullFilter+="$filter|"
        done

        filteringCmd="_filedir $fullFilter"
        __bb_debug "File filtering command: $filteringCmd"
        $filteringCmd
    elif (((directive & shellCompDirectiveFilterDirs) != 0)); then
        # File completion for directories only

        local subdir
        subdir=${completions[0]}
        if [[ -n $subdir ]]; then
            __bb_debug "Listing directories in $subdir"
            pushd "$subdir" >/dev/null 2>&1 && _filedir -d && popd >/dev/null 2>&1 || return
        else
            __bb_debug "Listing directories in ."
            _filedir -d
        fi
    else
        __bb_handle_completion_types
    fi

    __bb_handle_special_char "$cur" :
    __bb_handle_special_char "$cur" =

    # Print the activeHelp statements before we finish
    if ((${#activeHelp[*]} != 0)); then
        printf "\n";
        printf "%s\n" "${activeHelp[@]}"
        printf "\n"

        # The prompt format is only available from bash 4.4.
        # We test if it is available before using it.
        if (x=${PS1@P}) 2> /dev/null; then
            printf "%s" "${PS1@P}${COMP_LINE[@]}"
        else
            # Can't print the prompt.  Just print the
            # text the user had typed, it is workable enough.
            printf "%s" "${COMP_LINE[@]}"
        fi
    fi
}

# Separate activeHelp lines from real completions.
# Fills the $activeHelp and $completions arrays.
__bb_extract_activeHelp() {
    local activeHelpMarker="_activeHelp_ "
    local endIndex=${#activeHelpMarker}

    while IFS='' read -r comp; do
        if [[ ${comp:0:endIndex} == $activeHelpMarker ]]; then
            comp=${comp:endIndex}
            __bb_debug "ActiveHelp found: $comp"
            if [[ -n $comp ]]; then
                activeHelp+=("$comp")
            fi
        else
            # Not an activeHelp line but a normal completion
            completions+=("$comp")
        fi
    done <<<"${out}"
}

__bb_handle_completion_types() {
    __bb_debug "__bb_handle_completion_types: COMP_TYPE is $COMP_TYPE"

    case $COMP_TYPE in
    37|42)
        # Type: menu-complete/menu-complete-backward and insert-completions
        # If the user requested inserting one completion at a time, or all
        # completions at once on the command-line we must remove the descriptions.
        # https://github.com/spf13/cobra/issues/1508
        local tab=$'\t' comp
        while IFS='' read -r comp; do
            [[ -z $comp ]] && continue
            # Strip any description
            comp=${comp%%$tab*}
            # Only consider the completions that match
            if [[ $comp == "$cur"* ]]; then
                COMPREPLY+=("$comp")
            fi
        done < <(printf "%s\n" "${completions[@]}")
        ;;

    *)
        # Type: complete (normal completion)
        __bb_handle_standard_completion_case
        ;;
    esac
}

__bb_handle_standard_completion_case() {
    local tab=$'\t' comp

    # Short circuit to optimize if we don't have descriptions
    if [[ "${completions[*]}" != *$tab* ]]; then
        IFS=$'\n' read -ra COMPREPLY -d '' < <(compgen -W "${completions[*]}" -- "$cur")
        return 0
    fi

    local longest=0
    local compline
    # Look for the longest completion so that we can format things nicely
    while IFS='' read -r compline; do
        [[ -z $compline ]] && continue
        # Strip any description before checking the length
        comp=${compline%%$tab*}
        # Only consider the completions that match
        [[ $comp == "$cur"* ]] || continue
        COMPREPLY+=("$compline")
        if ((${#comp}>longest)); then
            longest=${#comp}
        fi
    done < <(printf "%s\n" "${completions[@]}")

    # If there is a single completion left, remove the description text
    if ((${#COMPREPLY[*]} == 1)); then
        __bb_debug "COMPREPLY[0]: ${COMPREPLY[0]}"
        comp="${COMPREPLY[0]%%$tab*}"
        __bb_debug "Removed description from single completion, which is now: ${comp}"
        COMPREPLY[0]=$comp
    else # Format the descriptions
        __bb_format_comp_descriptions $longest
    fi
}

__bb_handle_special_char()
{
    local comp="$1"
    local char=$2
    if [[ "$comp" == *${char}* && "$COMP_WORDBREAKS" == *${char}* ]]; then
        local word=${comp%"${comp##*${char}}"}
        local idx=${#COMPREPLY[*]}
        while ((--idx >= 0)); do
            COMPREPLY[idx]=${COMPREPLY[idx]#"$word"}
        done
    fi
}

__bb_format_comp_descriptions()
{
    local tab=$'\t'
    local comp desc maxdesclength
    local longest=$1

    local i ci
    for ci in ${!COMPREPLY[*]}; do
        comp=${COMPREPLY[ci]}
        # Properly format the description string which follows a tab character if there is one
        if [[ "$comp" == *$tab* ]]; then
            __bb_debug "Original comp: $comp"
            desc=${comp#*$tab}
            comp=${comp%%$tab*}

            # $COLUMNS stores the current shell width.
            # Remove an extra 4 because we add 2 spaces and 2 parentheses.
            maxdesclength=$(( COLUMNS - longest - 4 ))

            # Make sure we can fit a description of at least 8 characters
            # if we are to align the descriptions.
            if ((maxdesclength > 8)); then
                # Add the proper number of spaces to align the descriptions
                for ((i = ${#comp} ; i < longest ; i++)); do
                    comp+=" "
                done
            else
                # Don't pad the descriptions so we can fit more text after the completion
                maxdesclength=$(( COLUMNS - ${#comp} - 4 ))
            fi

            # If there is enough space for any description text,
            # truncate the descriptions that are too long for the shell width
            if ((maxdesclength > 0)); then
                if ((${#desc} > maxdesclength)); then
                    desc=${desc:0:$(( maxdesclength - 1 ))}
                    desc+="…"
                fi
                comp+="  ($desc)"
            fi
            COMPREPLY[ci]=$comp
            __bb_debug "Final comp: $comp"
        fi
    done
}

__start_bb()
{
    local cur prev words cword split

    COMPREPLY=()

    # Call _init_completion from the bash-completion package
    # to prepare the arguments properly
    if declare -F _init_completion >/dev/null 2>&1; then
        _init_completion -n =: || return
    else
        __bb_init_completion -n =: || return
    fi

    __bb_debug
    __bb_debug "========= starting completion logic =========="
    __bb_debug "cur is ${cur}, words[*] is ${words[*]}, #words[@] is ${#words[@]}, cword is $cword"

    # The user could have moved the cursor backwards on the command-line.
    # We need to trigger completion from the $cword location, so we need
    # to truncate the command-line ($words) up to the $cword location.
    words=("${words[@]:0:$cword+1}")
    __bb_debug "Truncated words[*]: ${words[*]},"

    local out directive
    __bb_get_completion_results
    __bb_process_completion_results
}

if [[ $(type -t compopt) = "builtin" ]]; then
    complete -o default -F __start_bb bb
else
    complete -o default -o nospace -F __start_bb bb
fi

# ex: ts=4 sw=4 et filetype=sh
```

</p>
</details> 

Once again thanks for your time and your awesome work !!
Cheers !!

**Comment by akinomyoga** [comment]
Thanks for the report. I have pushed a fix to the master. Could you check if the problem is fixed?

> Do you have any clue what's causing it ? I'm also going to attache the generated bash completion file in case it helps.

Thank you for attaching the generated script. The attached script suggests that the completion code is generated by the [`spf13/cobra`](https://github.com/spf13/cobra) library.

Bash doesn't support showing the description of each completion, but the `cobra` library exploits Bash's completion behavior where it does not insert completion texts when the programmable completion is attempted for showing menus. `cobra` includes the description as a part of the completed text. However, I must say this implementation is merely a hack and unstable. For example, it doesn't work with a certain combination of the readline settings. Nevertheless, ble.sh works around the problem by detecting `cobra` and overwriting a function defined by `cobra` to split and extract the generated description.

This time, the `cobra` completion seems to have been updated, and an introduced inconsistency between the base `cobra` completion and the overwriting function caused the problem. The `cobra` completion script was updated by https://github.com/spf13/cobra/pull/1482 ([`bash_completionsV2.go`](https://github.com/spf13/cobra/pull/1482/files#diff-ccbad943e447c0dbcf310892446819dea78bd471093c627e4aeb211c6ebb1b63)), and the change has been released in v1.5.0 (2022-06-21). I updated ble.sh so that it detects the version of the `cobra` completion by looking at the existence of the function `__{commandName}_extract_activeHelp`.

**Comment by 3ximus** [comment]
I see! Nothing could be done better on cobra's side ?
Glad that you noticed it right away though! Your patch does fix the issue!
I really appreciate it :+1: 
Cheers!
